File: golang/day2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	data := readFile("../../input/day2.txt")
	fmt.Println(Part1(data))
	fmt.Println(Part2(data))
}

func readFile(filename string) string {
	file, _ := os.Open(filename)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	if scanner.Scan() {
		return scanner.Text()
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}

	return ""

}

func Part1(s string) int {

	count := 0

	for rangeStr := range strings.SplitSeq(s, ",") {

		splitRange := strings.Split(rangeStr, "-")
		start, _ := strconv.Atoi(splitRange[0])
		end, _ := strconv.Atoi(splitRange[1])

		for i := start; i <= end; i++ {
			id := strconv.Itoa(i)
			mid := len(id) / 2
			if id[mid:] == id[:mid] {
				count += i
			}
		}
	}

	return count

}

func Part2(s string) int {

	count := 0

	for rangeStr := range strings.SplitSeq(s, ",") {

		start, _ := strconv.Atoi(strings.Split(rangeStr, "-")[0])
		end, _ := strconv.Atoi(strings.Split(rangeStr, "-")[1])

		for i := start; i <= end; i++ {
			id := strconv.Itoa(i)
			doubled := id + id
			if strings.Contains(doubled[1:len(doubled)-1], id) {
				count += i
			}
		}
	}

	return count

}

File: python/day2/main.py
def read_input(filename: str) -> str:
    with open(filename, "r", encoding="utf-8") as f:
        return f.read()


inputPath = "../../input/day2.txt"


def part1() -> int:
    input = read_input(inputPath)
    invalid_ids_sum: int = 0

    for range_str in input.split(","):
        start_str, end_str = range_str.split("-")
        start, end = int(start_str), int(end_str)

        for r in range(start, end + 1):
            id = str(r)
            mid = len(id) // 2
            if id[:mid] == id[mid:]:
                invalid_ids_sum += r

    return invalid_ids_sum


def part2() -> int:
    input = read_input(inputPath)
    invalid_ids_sum: int = 0

    for range_str in input.split(","):
        start_str, end_str = range_str.split("-")
        start, end = int(start_str), int(end_str)

        for r in range(start, end + 1):
            id = str(r)
            if id in (id + id)[1:-1]:
                invalid_ids_sum += r

    return invalid_ids_sum


if __name__ == "__main__":
    print(part1())
    print(part2())
